Mypy disallows overriding a generic type attribute with a property

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

If I have a generic class with an attribute of a generic type, when creating a concrete subclass, overriding that attribute with a correctly typed property raises an error. Similar to #4125

**To Reproduce**

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class A(Generic[T]):
    def __init__(self, val: T):
        self.member: T = val

class B(A[str]):
    @property
    def member(self) -> str:
        return "bar"
    
    @member.setter
    def member(self, val: str):
        pass
```

**Actual Behavior**

```
property_issue.py:10: error: Signature of "member" incompatible with supertype "A"  [override]
property_issue.py:10: note:      Superclass:
property_issue.py:10: note:          T
property_issue.py:10: note:      Subclass:
property_issue.py:10: note:          str
```
<!-- What went wrong? Paste mypy's output. -->

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 1.9.0
- Mypy command-line flags: -
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.9.13

<!-- You can freely edit this text, please remove all the lines you believe are unnecessary. -->

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -2095,6 +2095,13 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 if original_node and is_property(original_node):
                     original_type = get_property_type(original_type)
 
+            if isinstance(original_node, Var):
+                expanded_type = map_type_from_supertype(original_type, defn.info, base)
+                expanded_type = expand_self_type(
+                    original_node, expanded_type, fill_typevars(defn.info)
+                )
+                original_type = get_proper_type(expanded_type)
+
             if is_property(defn):
                 inner: FunctionLike | None
                 if isinstance(typ, FunctionLike):

--- a/test-data/unit/check-generic-subtyping.test
+++ b/test-data/unit/check-generic-subtyping.test
@@ -1108,6 +1108,28 @@ class BadGenericPropertyOverride(A[str], Generic[T]):
     def member(self, val: T): ...
 [builtins fixtures/property.pyi]
 
+[case testParameterizedGenericOverrideSelfWithProperty]
+from typing_extensions import Self
+
+class A:
+    def __init__(self, val: Self):
+        self.member: Self = val
+
+class GoodPropertyOverride(A):
+    @property
+    def member(self) -> "GoodPropertyOverride": ...
+
+    @member.setter
+    def member(self, val: "GoodPropertyOverride"): ...
+
+class GoodPropertyOverrideSelf(A):
+    @property
+    def member(self) -> Self: ...
+
+    @member.setter
+    def member(self, val: Self): ...
+[builtins fixtures/property.pyi]
+
 [case testParameterizedGenericOverrideWithSelfProperty]
 from typing import TypeVar, Generic
 from typing_extensions import Self

--- a/test-data/unit/check-generic-subtyping.test
+++ b/test-data/unit/check-generic-subtyping.test
@@ -1081,7 +1081,6 @@ class B(A[str]):
 class GoodPropertyOverride(A[str]):
     @property
     def member(self) -> str: ...
-
     @member.setter
     def member(self, val: str): ...
 
@@ -1092,7 +1091,6 @@ class BadPropertyOverride(A[str]):
                # N:      Subclass: \
                # N:          int
     def member(self) -> int: ...
-
     @member.setter
     def member(self, val: int): ...
 
@@ -1103,7 +1101,47 @@ class BadGenericPropertyOverride(A[str], Generic[T]):
                # N:      Subclass: \
                # N:          T
     def member(self) -> T: ...
+    @member.setter
+    def member(self, val: T): ...
+[builtins fixtures/property.pyi]
 
+[case testParameterizedGenericPropertyOverrideWithProperty]
+from typing import TypeVar, Generic
+
+T = TypeVar("T")
+
+class A(Generic[T]):
+    @property
+    def member(self) -> T: ...
+    @member.setter
+    def member(self, val: T): ...
+
+class B(A[str]):
+    member: str
+
+class GoodPropertyOverride(A[str]):
+    @property
+    def member(self) -> str: ...
+    @member.setter
+    def member(self, val: str): ...
+
+class BadPropertyOverride(A[str]):
+    @property  # E: Signature of "member" incompatible with supertype "A" \
+               # N:      Superclass: \
+               # N:          str \
+               # N:      Subclass: \
+               # N:          int
+    def member(self) -> int: ...
+    @member.setter
+    def member(self, val: int): ...
+
+class BadGenericPropertyOverride(A[str], Generic[T]):
+    @property  # E: Signature of "member" incompatible with supertype "A" \
+               # N:      Superclass: \
+               # N:          str \
+               # N:      Subclass: \
+               # N:          T
+    def member(self) -> T: ...
     @member.setter
     def member(self, val: T): ...
 [builtins fixtures/property.pyi]
@@ -1118,14 +1156,12 @@ class A:
 class GoodPropertyOverride(A):
     @property
     def member(self) -> "GoodPropertyOverride": ...
-
     @member.setter
     def member(self, val: "GoodPropertyOverride"): ...
 
 class GoodPropertyOverrideSelf(A):
     @property
     def member(self) -> Self: ...
-
     @member.setter
     def member(self, val: Self): ...
 [builtins fixtures/property.pyi]
@@ -1146,7 +1182,6 @@ class B(A["B"]):
 class GoodPropertyOverride(A["GoodPropertyOverride"]):
     @property
     def member(self) -> Self: ...
-
     @member.setter
     def member(self, val: Self): ...
 [builtins fixtures/property.pyi]

--- a/test-data/unit/check-generic-subtyping.test
+++ b/test-data/unit/check-generic-subtyping.test
@@ -1065,3 +1065,66 @@ class F(E[T_co], Generic[T_co]): ...  # E: Variance of TypeVar "T_co" incompatib
 
 class G(Generic[T]): ...
 class H(G[T_contra], Generic[T_contra]): ...  # E: Variance of TypeVar "T_contra" incompatible with variance in parent type
+
+[case testParameterizedGenericOverrideWithProperty]
+from typing import TypeVar, Generic
+
+T = TypeVar("T")
+
+class A(Generic[T]):
+    def __init__(self, val: T):
+        self.member: T = val
+
+class B(A[str]):
+    member: str
+
+class GoodPropertyOverride(A[str]):
+    @property
+    def member(self) -> str: ...
+
+    @member.setter
+    def member(self, val: str): ...
+
+class BadPropertyOverride(A[str]):
+    @property  # E: Signature of "member" incompatible with supertype "A" \
+               # N:      Superclass: \
+               # N:          str \
+               # N:      Subclass: \
+               # N:          int
+    def member(self) -> int: ...
+
+    @member.setter
+    def member(self, val: int): ...
+
+class BadGenericPropertyOverride(A[str], Generic[T]):
+    @property  # E: Signature of "member" incompatible with supertype "A" \
+               # N:      Superclass: \
+               # N:          str \
+               # N:      Subclass: \
+               # N:          T
+    def member(self) -> T: ...
+
+    @member.setter
+    def member(self, val: T): ...
+[builtins fixtures/property.pyi]
+
+[case testParameterizedGenericOverrideWithSelfProperty]
+from typing import TypeVar, Generic
+from typing_extensions import Self
+
+T = TypeVar("T")
+
+class A(Generic[T]):
+    def __init__(self, val: T):
+        self.member: T = val
+
+class B(A["B"]):
+    member: Self
+
+class GoodPropertyOverride(A["GoodPropertyOverride"]):
+    @property
+    def member(self) -> Self: ...
+
+    @member.setter
+    def member(self, val: Self): ...
+[builtins fixtures/property.pyi]

